daiquiri/formatter.py

```diff
@@ -23,19 +31,124 @@
 )
 
 
-class ColorFormatter(logging.Formatter):
-    # TODO(jd) Allow configuration
-    LEVEL_COLORS = {
-        logging.DEBUG: '\033[00;32m',  # GREEN
-        logging.INFO: '\033[00;36m',  # CYAN
-        logging.WARN: '\033[01;33m',  # BOLD YELLOW
-        logging.ERROR: '\033[01;31m',  # BOLD RED
-        logging.CRITICAL: '\033[01;31m',  # BOLD RED
-    }
+class Palette(object):
+    """Object that represents TTY colors."""
+
+    PREFIX = '\033['
+    BASE_WHITE = 15
+    BASE_MOD = '00;%dm'
+    CHART_WHITE = 231
+    CHART_MOD = '38;5;%dm'
+
+    def __init__(self, *args, base=False, name=None):
+        """Create a new color palette from console values.
+
+        :param args: a color value for each of the default logging levels.
+        :param base: (bool) use TTY defined colors instead of 256-color palette.
+        """
+        self.name = name
+        self.base = base
+        self.colors = list(args)
+        self._level_colors = None
+
+    def __str__(self):
+        colors = [c[1] if not isinstance(c, int) else c for c in self.colors]
+        return '<Palette %s(%s)>' % (self.name if self.name else '',
+                                     ', '.join(map(str, colors)))
+
+    @property
+    def level_colors(self):
+        # deferred processing, read only
+        if self._level_colors is None:
+            self._level_colors = self._process_colors()
+        return self._level_colors
+
+    def _process_colors(self):
+        ret = {}
+        # account for custom logging levels, with potential additional colors
+        logging_levels = tuple(sorted(logging._levelToName.keys(), reverse=False))
+
+        # use `white` for all levels that extend beyond the palette's values
+        white = self.BASE_WHITE if self.base else self.CHART_WHITE
+
+        # create a tuple pair for each level and the palette's colors
+        colors = zipl(logging_levels, [white] + self.colors, fillvalue=white)
+
+        # apply text formatting
+        for level, color in colors:
+            if isinstance(color, tuple):
+                # text-decoration modifier
+                fmt, c_num = color
+            elif isinstance(color, int):
+                # no formatting
+                fmt, c_num = normal(color)
+            else:
+                raise ValueError(color)
+
+            if self.base:
+                c = self.BASE_MOD % c_num
+            else:
+                c = self.CHART_MOD % c_num
+            ret[level] = '%s%s' % (self.PREFIX, fmt % c)
+        return ret
+
+
+def bold(n):
+    """Bold color in Palette.
 
+    :param n: (int)
+    """
+    return '1;%s', n
+
+
+def normal(n):
+    """Normal, no decorations, color in Palette.
+
+    :param n: (int)
+    """
+    return '0;%s', n
+
+
+def underline(n):
```

Comment:
No clue how to use this function. I imagine it should also be attached to a class? Even as a static method.